zcash_client_sqlite/src/wallet.rs

```diff
@@ -2311,6 +2314,371 @@ pub(crate) fn put_block(
+                    let wallet_address =
+                        select_receiving_address(params, conn, *output.account(), &receiver)?
+                            .unwrap_or_else(|| receiver.to_zcash_address(params.network_type()));
+
+                    Recipient::External(wallet_address, PoolType::ORCHARD)
+                };
+
+                put_sent_output(
+                    conn,
+                    params,
+                    *output.account(),
+                    tx_ref,
+                    output.index(),
+                    &recipient,
+                    output.note_value(),
+                    Some(output.memo()),
+                )?;
+            }
+            TransferType::WalletInternal => {
+                orchard::put_received_note(conn, output, tx_ref, None)?;
+
+                let recipient = Recipient::InternalAccount {
+                    receiving_account: *output.account(),
+                    external_address: None,
+                    note: Note::Orchard(*output.note()),
+                };
+
+                put_sent_output(
+                    conn,
+                    params,
+                    *output.account(),
+                    tx_ref,
+                    output.index(),
+                    &recipient,
+                    output.note_value(),
+                    Some(output.memo()),
+                )?;
+            }
+            TransferType::Incoming => {
+                orchard::put_received_note(conn, output, tx_ref, None)?;
+
+                if let Some(account_id) = funding_account {
+                    // Even if the recipient address is external, record the send as internal.
+                    let recipient = Recipient::InternalAccount {
+                        receiving_account: *output.account(),
+                        external_address: {
+                            let receiver = Receiver::Orchard(output.note().recipient());
+                            Some(
+                                select_receiving_address(
+                                    params,
+                                    conn,
+                                    *output.account(),
+                                    &receiver,
+                                )?
+                                .unwrap_or_else(|| {
+                                    receiver.to_zcash_address(params.network_type())
+                                }),
+                            )
+                        },
+                        note: Note::Orchard(*output.note()),
+                    };
+
+                    put_sent_output(
+                        conn,
+                        params,
+                        account_id,
+                        tx_ref,
+                        output.index(),
+                        &recipient,
+                        output.note_value(),
+                        Some(output.memo()),
+                    )?;
+                }
+            }
+        }
```

Comment:
This method is way too long. How about factoring out this to `mark_orchard_output_spent`, analogous to `transparent::mark_transparent_utxo_spent`? (And similarly for Sapling.)

Reply:
I agree that this could be factored out, although `mark_orchard_output_spent` is not at all the semantics. However, I don't think we need to worry about that now.

Reply:
In any case, this just shows up as new here because of the move in  https://github.com/zcash/librustzcash/pull/1496/commits/75f236a148c936c34bfaa96f0986780695b7a231 - it shouldn't be moved and refactored at once.